### ISSUE TITLE
Hundreds of libpng warnings when compiling local tiles

### ISSUE THREAD

**Issue by randomMesh** [crawl/crawl#1710]
**PLATFORM and OS or BROWSER:**
- Local tiles
- Debian GNU/Linux 10
- GCC 8.3.0
- libpng-dev 1.6.36

**DCSS VERSION AND GAME INFORMATION:**
Latest master branch

**DESCRIPTION:**
Compiling local tiles from the latest master branch spams the console with a lot of libpng warnings:
```bash
xx@xxx:~/development/crawl/crawl-ref/source$ make TILES=y
util/art-data.pl
    Generating art-data.h
    Generating art-enum.h
    Generating dc-unrand.txt
    Generating tiledef-unrand.cc
make -C rltiles all ARCH=i686-linux-gnu NO_PKGCONFIG= TILES=y
make[1]: Verzeichnis „/home/xxx/development/crawl/crawl-ref/source/rltiles“ wird betreten
    GEN main.png
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
    GEN tiledef-main.h
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
    GEN player.png
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
    GEN tiledef-player.h
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: Interlace handling should be turned on when using png_read_image
libpng warning: iCCP: known incorrect sRGB profile
    GEN gui.png

<- snip ->
```
While not a critical bug since DCSS is compiled successfully, it's ugly as hell and may scare off potential collaborators.

**Comment by rawlins** [comment]
I won't immediately close this but this is most likely wontfix in the sense that I doubt we can or will fix the specific issue ourself; this is a libpng issue, not a dcss issue, and it's a harmless and super common one with libpng on top of that. See e.g. https://sourceforge.net/p/libpng/bugs/165/ for some details; supposedly fixed at some point, but apparently still comes and goes, based on searching for the error. It's possible updating contribs would address this, but not certain (and that has complications as well).

>it's ugly as hell and may scare off potential collaborators.

If someone is scared by this, I'm not sure what crawlcode itself will do to them.

**Comment by randomMesh** [comment]
One possibility would be do de-interlace all these images. ImageMagicks `convert` can do that.

**Comment by Aliscans** [comment]
For what it's worth, running every PNG through imagemagick does remove those errors, as well as the "iCCP: known incorrect sRGB profile" ones.

It is overkill, though, as there are only around 100 images which trigger either warning.

**Comment by randomMesh** [comment]
The warnings would be gone, the overall size of DCSS would be smaller and it would load faster. I can only see advantages.
The conversion process can be automated, so this should not take a long time.

**Comment by ebering** [comment]
Tile optimization is included in our build process, using `advpng` and `pngcrush`. The assets shipped with a build are the composed tilesheets compiled by rltiles, and optimizing these at build time will be necessary whether or not the repo tiles are regularly optimized by a service.

A PR to fix interlacing in those pngs that generate warnings would be welcome, but as @rawlins said its not a real priority.

Imgbot is probably more of a headache than a benefit for us. The pngs in rltiles aren't a great compression headache: they're small to begin with, most have been optimized at some point in their lifetime, and they get re-compressed with the build.

**Comment by randomMesh** [comment]
> A PR to fix interlacing in those pngs that generate warnings would be welcome

I just wrote a script and did the conversion. Looks good, no more warnings. If my tests are successful, i'll open a PR.

**Comment by rawlins** [comment]
> You're right, the code could use a makeup. It's nice that you guys add features, remove features, fix bugs. But working on readability is also important as otherwise this can get quickly out of hand.
> 
> Any plans to modernize the codebase in the future?

Glad you're interested in chatting about crawl. Please know that crawlcode got out of hand right from the start, is a large codebase with many contributors going back to 1999, and we in fact do expend quite a bit of effort on readability and code improvement already. It's big enough that it isn't really a matter of just simply sitting down and "modernizing the codebase"! Readability patches are welcome; convention docs [here](https://github.com/crawl/crawl/blob/master/crawl-ref/docs/develop/coding_conventions.txt) and starting contributor guide [here](https://github.com/crawl/crawl/blob/master/crawl-ref/docs/develop/contribution-process.md).

**Comment by randomMesh** [comment]
I've ran my script over all the images and there are 2767 images which are either interlaced, too large or corrupt.
I've logged it but the log is ~50MB in size, so i won't post it. Here's the first entry as an example:

```bash
** Processing: /home/xxx/development/crawl/crawl-ref/source/rltiles/player/hand2/kite_shield_of_resistance.png
Warning: Extraneous data found after IEND
32x32 pixels, 4x8 bits/pixel, RGB+alpha
Reducing image to 4 bits/pixel, 8 colors (1 transparent) in palette
Recoverable errors found in input. Fixing...
Input IDAT size = 262 bytes
Input file size = 319 bytes

Trying:
  zc = 9  zm = 9  zs = 0  f = 0		IDAT size = 106
  zc = 9  zm = 8  zs = 0  f = 0		IDAT size = 106
  zc = 8  zm = 9  zs = 0  f = 0		IDAT size = 106
  zc = 8  zm = 8  zs = 0  f = 0		IDAT size = 106
  zc = 7  zm = 9  zs = 0  f = 0		IDAT size = 106
  zc = 7  zm = 8  zs = 0  f = 0		IDAT size = 106
  zc = 6  zm = 9  zs = 0  f = 0		IDAT size = 106
  zc = 6  zm = 8  zs = 0  f = 0		IDAT size = 106
  zc = 5  zm = 8  zs = 0  f = 0		IDAT size = 106
  zc = 4  zm = 8  zs = 0  f = 0		IDAT size = 106
  zc = 9  zm = 9  zs = 1  f = 0		IDAT size = 106
  zc = 9  zm = 8  zs = 1  f = 0		IDAT size = 106
  zc = 8  zm = 9  zs = 1  f = 0		IDAT size = 106
  zc = 8  zm = 8  zs = 1  f = 0		IDAT size = 106
  zc = 7  zm = 9  zs = 1  f = 0		IDAT size = 106
  zc = 7  zm = 8  zs = 1  f = 0		IDAT size = 106
  zc = 6  zm = 9  zs = 1  f = 0		IDAT size = 106
  zc = 6  zm = 8  zs = 1  f = 0		IDAT size = 106
  zc = 5  zm = 8  zs = 1  f = 0		IDAT size = 106
  zc = 4  zm = 8  zs = 1  f = 0		IDAT size = 106
                               
Selecting parameters:
  zc = 4  zm = 8  zs = 1  f = 0		IDAT size = 106

Output IDAT size = 106 bytes (156 bytes decrease)
Output file size = 212 bytes (107 bytes = 33.54% decrease)

** Status report
1 file(s) have been processed.
1 error(s) have been encountered.
1 erroneous file(s) have been fixed.
```

@ebering : How should the PR look like? Do you want all the 2767 images or just the script?

**Comment by ebering** [comment]
> > A PR to fix interlacing in those pngs that generate warnings would be welcome
> 
> @ebering : How should the PR look like? Do you want all the 2767 images or just the script?
> The script does nothing special, just calls OptiPNG in a loop with logging to a file.

The script, please! Place it in `crawl-ref/source/util`, with a name like checkpng. If it isn't too much trouble, please make it work like our linting programs checkwhite and unbrace by supporting the following command line switches:
- `-n` for no change, only report 
-  `-m` for only checking modified files in the git index

I hadn't been aware that we had so many images with issues, so I've had a change of heart about whether or not we need regular checking—however I'd prefer an in-house solution to something like Imagebot.

**Comment by randomMesh** [comment]
I've created a pull request for the script: #1727

The -n parameter works but the -m parameter is not implemented yet.

Instructions to run the script are in the comments in that file.
Please let me know if something is not clear or should be changed.

**Comment by rawlins** [comment]
Should be resolved by ab5fb1643382
